Implement and use cuda graph plans

wishstudio opened the pull request:
This PR tries to implement graph plan APIs for the CUDA backend, as well as implement code in ggml-backend.cpp to actually use the graph plan APIs when a backend supports it.

The main functional improvement is to support cuda graphs when the graph is split (e.g. for hybrid inference). Currently the graph update and reuse logic (`ggml_backend_sched_update_plans`) is a simple heuristic: only try updating previous graphs when the number of splits and their corresponding backends are the same as the previous run. As the benchmark shown this universally accelerate hybrid inference tg performance by up to 30%.

The CUDA graph execution code is refactored and cleaned up. Two out of three original graph plan fail path are removed: `disable_due_to_failed_graph_capture` and `disable_due_to_too_many_updates`. The former one is due to the fact I found no code setting it to true. The latter is because I have currently no idea about the semantics in a split graph scenario. But it seems to not degrade the performance at all. Interestingly, I found that on my rig, even repeatitively build a graph then execute it only once is always faster than calling the kernels individually. I suspect it is the reason that the performance increased in tests even for CUDA only workloads, given this PR's optimization not targeting them. This of course needs to be verified on more hardware configurations.

Performance comparison:
RTX 5090 + 13700k, 128GB 6400 MT/s RAM
| model | n_cpu_moe | test | t/s master | t/s pr | Speedup |
|-|-|-|-|-|-|
| gpt-oss 20B MXFP4 MoE | 0 | pp512 | 9070.14 | 8768.06 | 0.97 |
| gpt-oss 20B MXFP4 MoE | 0 | tg128 | 273.99 | 278.43 | 1.02 |
| gpt-oss 20B MXFP4 MoE | 99 | pp512 | 916.16 | 931.72 | 1.02 |
| gpt-oss 20B MXFP4 MoE | 99 | tg128 | 42.4 | 47.2 | 1.11 |
| gpt-oss 120B MXFP4 MoE | 24 | pp512 | 150.76 | 150.31 | 1.00 |
| gpt-oss 120B MXFP4 MoE | 24 | tg128 | 36.73 | 45.04 | 1.23 |
| gpt-oss 120B MXFP4 MoE | 99 | pp512 | 187.69 | 186.21 | 0.99 |
| gpt-oss 120B MXFP4 MoE | 99 | tg128 | 28.24 | 31.7 | 1.12 |
| glm4moe 106B.A12B Q4_K | 34 | pp512 | 81.4 | 79.9 | 0.98 |
| glm4moe 106B.A12B Q4_K | 34 | tg128 | 18.69 | 21.72 | 1.16 |
| glm4moe 106B.A12B Q4_K | 99 | pp512 | 114.85 | 114.34 | 1.00 |
| glm4moe 106B.A12B Q4_K | 99 | tg128 | 14.59 | 16.01 | 1.10 |
| glm4moe 355B.A32B Q2_K | 99 | pp512 | 25.3 | 26.96 | 1.07 |
| glm4moe 355B.A32B Q2_K | 99 | tg128 | 7.73 | 8.74 | 1.13 |
| qwen3moe 235B.A22B Q3_K | 80 | pp512 | 59.99 | 61.26 | 1.02 |
| qwen3moe 235B.A22B Q3_K | 80 | tg128 | 10.66 | 11.77 | 1.10 |
| qwen3moe 235B.A22B Q3_K | 99 | pp512 | 85.38 | 88.45 | 1.04 |
| qwen3moe 235B.A22B Q3_K | 99 | tg128 | 9.27 | 10.21 | 1.10 |
| qwen3moe 30B.A3B Q4_K | 0 | pp512 | 6806.34 | 7295.45 | 1.07 |
| qwen3moe 30B.A3B Q4_K | 0 | tg128 | 246.12 | 273 | 1.11 |
| qwen3moe 30B.A3B Q4_K | 99 | pp512 | 531.72 | 560.52 | 1.05 |
| qwen3moe 30B.A3B Q4_K | 99 | tg128 | 36.1 | 48.4 | 1.34 |
| qwen3 8B Q8_0 |  | pp512 | 10002.11 | 11383.42 | 1.14 |
| qwen3 8B Q8_0 |  | tg128 | 124.07 | 136.75 | 1.10 |
| llama 13B Q6_K |  | pp512 | 3886.45 | 4175.62 | 1.07 |
| llama 13B Q6_K |  | tg128 | 65.89 | 72.19 | 1.10 |
| llama 8B Q8_0 |  | pp512 | 10092.17 | 11702.73 | 1.16 |
| llama 8B Q8_0 |  | tg128 | 127.96 | 142.67 | 1.11 |
| gemma3 12B Q8_0 |  | pp512 | 6696.89 | 7920.99 | 1.18 |
| gemma3 12B Q8_0 |  | tg128 | 79.01 | 88.73 | 1.12 |
| nemotron_h 9B Q8_0 |  | pp512 | 7243.57 | 7808.74 | 1.08 |
| nemotron_h 9B Q8_0 |  | tg128 | 114.51 | 122.93 | 1.07 |

wishstudio committed:
Implement and use cuda graph plans.

wishstudio committed:
Merge remote-tracking branch 'upstream' into cuda_graph_plan

IMbackK commented:
One thing to mention here is that on hip the generation of the graph is relatively more expensive than on cuda which makes the removal disable_due_to_too_many_updates have an perhaps outsized negative effect there. I will try to find the time to investigate.

wishstudio commented:
@AesSedai Could you give a try at gpt-oss-120b? I found that for some reason this PR applied to latest master works much slower for GLM 4.5 Air. Will need to investigate further. My testing shows gpt-oss-120b still has the same speedup.

@IMbackK My wild guess is that if the graph gets reused for a few times it should pay off the graph building overhead. Without this PR it is understandable to have this `disable_due_to_too_many_updates` condition because in a split graph nothing will get reused at all. But after this PR I'm curious at on what scenario are we expected to see a very high update frequency.

AesSedai commented:
I'll have to download it and set it up, but I should be able to run it later tonight.

wishstudio committed:
Merge branch 'master' into cuda_graph_plan

DocShotgun commented:
Ran some sweeps on my system for CPU+GPU inference - 5th gen Xeon with 8ch DDR5-5600 (it's a dual socket server but I bound to a single socket due to llama.cpp not having NUMA parallelism) + RTX PRO 6000. Unfortunately appears to be non-beneficial on my setup (maybe very minmally detrimental).

`~/numactl-bind-socket.sh --socket 0 --mode all ~/llama.cpp/build/bin/llama-sweep-bench -m /mnt/data/models/GGUF/GLM-4.6-Q8_0-Q4_K-Q4_K-Q5_K.gguf -c 32768 -ngl 999 -fa on -t 64 -b 4096 -ub 4096 --no-mmap --numa numactl -ot "blk\.(3[2-9]|[4-9][0-9])\.ffn_.*_exps.=CPU"`

<img width="3404" height="1694" alt="thud-sweep" src="https://github.com/user-attachments/assets/751012bf-55d5-4a6a-8238-c32f66894e3f" />

```
## llama.cpp [master, 7adc79c03]
|    PP |     TG |   N_KV |   T_PP s | S_PP t/s |   T_TG s | S_TG t/s |
|-------|--------|--------|----------|----------|----------|----------|
|  4096 |   1024 |      0 |    4.627 |   885.15 |   83.217 |    12.31 |
|  4096 |   1024 |   4096 |    4.941 |   829.05 |   86.452 |    11.84 |
|  4096 |   1024 |   8192 |    5.259 |   778.81 |   87.578 |    11.69 |
|  4096 |   1024 |  12288 |    5.647 |   725.35 |   89.944 |    11.38 |
|  4096 |   1024 |  16384 |    5.970 |   686.06 |   91.353 |    11.21 |
|  4096 |   1024 |  20480 |    6.319 |   648.22 |   93.331 |    10.97 |
|  4096 |   1024 |  24576 |    6.682 |   612.98 |   95.518 |    10.72 |
|  4096 |   1024 |  28672 |    7.111 |   575.97 |   96.989 |    10.56 |

## llama.cpp [cuda_graph_plan, c17f8b5bd]
|    PP |     TG |   N_KV |   T_PP s | S_PP t/s |   T_TG s | S_TG t/s |
|-------|--------|--------|----------|----------|----------|----------|
|  4096 |   1024 |      0 |    4.633 |   884.15 |   84.082 |    12.18 |
|  4096 |   1024 |   4096 |    4.939 |   829.39 |   87.077 |    11.76 |
|  4096 |   1024 |   8192 |    5.264 |   778.08 |   88.174 |    11.61 |
|  4096 |   1024 |  12288 |    5.620 |   728.83 |   90.342 |    11.33 |
|  4096 |   1024 |  16384 |    5.965 |   686.63 |   91.545 |    11.19 |
|  4096 |   1024 |  20480 |    6.313 |   648.82 |   93.665 |    10.93 |
|  4096 |   1024 |  24576 |    6.681 |   613.04 |   95.739 |    10.70 |
|  4096 |   1024 |  28672 |    7.115 |   575.69 |   97.026 |    10.55 |
```

wishstudio committed:
Fix cuda graph update logic.

wishstudio commented:
Looks like I screwed up the previous commits as the the graph update logic is completely broken. I can't even reproduce my previous results :(

Anyway the issues should now be fixed. Refactored a bit and added back the `disable_due_to_too_many_updates` condition. It currently works at the cuda graph level that if a single cuda graph changes too frequently we disable all future graph creation. It is not identical to previous behavior in non split graphs but should be close.

@AesSedai @DocShotgun Thank you for your tests! Please check if this commit (3afbd9f) works.

<details>
<summary>Performance comparison</summary>

| model | n_cpu_moe | test | t/s master (7adc79c) | t/s pr (3afbd9f) | Speedup |
| -- | -- | -- | -- | -- | -- |
| gpt-oss 20B MXFP4 MoE | 0 | pp512 | 9927.91 | 10180.03 | 1.03 |
| gpt-oss 20B MXFP4 MoE | 0 | tg128 | 297.95 | 298.2 | 1.00 |
| gpt-oss 20B MXFP4 MoE | 99 | pp512 | 987.16 | 921.35 | 0.93 |
| gpt-oss 20B MXFP4 MoE | 99 | tg128 | 42.14 | 47.47 | 1.13 |
| gpt-oss 120B MXFP4 MoE | 24 | pp512 | 305.3 | 282.15 | 0.92 |
| gpt-oss 120B MXFP4 MoE | 24 | tg128 | 37.59 | 46.05 | 1.23 |
| gpt-oss 120B MXFP4 MoE | 99 | pp512 | 216.95 | 202.49 | 0.93 |
| gpt-oss 120B MXFP4 MoE | 99 | tg128 | 28.58 | 31.93 | 1.12 |
| glm4moe 106B.A12B Q4_K | 34 | pp512 | 166.5 | 147.04 | 0.88 |
| glm4moe 106B.A12B Q4_K | 34 | tg128 | 19.27 | 22.38 | 1.16 |
| glm4moe 106B.A12B Q4_K | 99 | pp512 | 132.63 | 118.56 | 0.89 |
| glm4moe 106B.A12B Q4_K | 99 | tg128 | 14.95 | 16.4 | 1.10 |
| glm4moe 355B.A32B Q2_K | 99 | pp512 | 44.17 | 46.51 | 1.05 |
| glm4moe 355B.A32B Q2_K | 99 | tg128 | 7.32 | 9.17 | 1.25 |
| qwen3moe 235B.A22B Q3_K | 80 | pp512 | 106.98 | 107.55 | 1.01 |
| qwen3moe 235B.A22B Q3_K | 80 | tg128 | 10.82 | 11.84 | 1.09 |
| qwen3moe 235B.A22B Q3_K | 99 | pp512 | 89.78 | 93.14 | 1.04 |
| qwen3moe 235B.A22B Q3_K | 99 | tg128 | 9.61 | 10.26 | 1.07 |
| qwen3moe 30B.A3B Q4_K | 0 | pp512 | 7260.62 | 7336.87 | 1.01 |
| qwen3moe 30B.A3B Q4_K | 0 | tg128 | 267.7 | 276.21 | 1.03 |
| qwen3moe 30B.A3B Q4_K | 99 | pp512 | 564.09 | 579.92 | 1.03 |
| qwen3moe 30B.A3B Q4_K | 99 | tg128 | 37.18 | 48.29 | 1.30 |
| qwen3 8B Q8_0 |  | pp512 | 10786.99 | 11638.03 | 1.08 |
| qwen3 8B Q8_0 |  | tg128 | 136.15 | 136.59 | 1.00 |
| llama 13B Q6_K |  | pp512 | 4208.3 | 4162.58 | 0.99 |
| llama 13B Q6_K |  | tg128 | 72.28 | 71.62 | 0.99 |
| llama 8B Q8_0 |  | pp512 | 10505.52 | 11916.53 | 1.13 |
| llama 8B Q8_0 |  | tg128 | 140.28 | 141.31 | 1.01 |
| gemma3 12B Q8_0 |  | pp512 | 7688.6 | 7593.39 | 0.99 |
| gemma3 12B Q8_0 |  | tg128 | 88.47 | 88.22 | 1.00 |
| nemotron_h 9B Q8_0 |  | pp512 | 8032.68 | 7850.61 | 0.98 |
| nemotron_h 9B Q8_0 |  | tg128 | 132.5 | 131.67 | 0.99 |
</details>

aendk commented:
Super interesting PR!

Off-topic to this PR itself, but relevant to the discussion: What is `llama-sweep-bench`? How do I build it?
I found nothing about it in the `master` branch when building by default, and nothing online.

ORippler commented:
> @DocShotgun @AesSedai @ubergarm Looks like all of you are running this in Linux and I'm running Win11. PP performance is very unstable in my tests but it looks much more stable on your tests. During my tests I also see lots of random latencies and slowdowns during kernel launches. This could explain that on my system this PR removes much more overhead than on your systems.

I can confirm that kernel launch overheads are significantly more expensive and varying on Win11 compared to Linux for NVGPUS due to their batched dispatch via WDDM2.

DocShotgun commented:
I've heard the same thing about kernel launch overhead being significantly higher on Windows than Linux. I notice you mention in the OP that you found building a graph and executing it only once to be always faster than calling the kernels individually - perhaps the lower kernel launch overhead on Linux makes this not necessarily true there?

wishstudio commented:
> I can confirm that kernel launch overheads are significantly more expensive and varying on Win11 compared to Linux for NVGPUS due to their batched dispatch via WDDM2.

Another reason to switch to Linux 🥲 

> I've heard the same thing about kernel launch overhead being significantly higher on Windows than Linux. I notice you mention in the OP that you found building a graph and executing it only once to be always faster than calling the kernels individually - perhaps the lower kernel launch overhead on Linux makes this not necessarily true there?

Could be. But in the last commit I introduced back the `disable_due_to_too_many_updates` guard that graph creation will be completely disabled after 4 consecutive graph updates. So unless it is a weird change pattern (like the graph updates 1-3 times and then gets reused 1 time, repetitively) it should no longer apply. In my tests with a lower quant of GLM 4.6 I can see the graphs are getting reused most of the time, updating only once every ~128 tokens. Another possibility if (overhead of graph manipulation) / 128 > (overhead reduction of graph launch).

ORippler reviewed (changes requested):
While I welcome the proposal, I feel there are issues that still need to be addressed, namely:
1. stricter equivalence checking when updating graph plans, and
2. fixing the logic in the cuda backend

---

To sched some light on the perf gains here some profiles (with locally hacked fix of point 2 raised above), where I offloaded a single FFN with

```
sudo nsys profile --output=gp_offload --trace=cuda,nvtx,osrt --sample=cpu --gpu-metrics-devices=all --cuda-graph-trace=node ./build-x64-linux-gcc-reldbg/bin/llama-cli -m models/gemma-3n-E2B-it-Q8_0.gguf -fa 1 -p "Hello World" --top-k 1 -ot "blk\.1.ffn*=CPU" --no-conversation -n 10
```

<img width="2275" height="1038" alt="image" src="https://github.com/user-attachments/assets/e7b2b166-8a5a-42f1-9b61-d5ac0330c7bd" />

One can see that CUDA Graphs nicely speed up the GPU portion of the workload (kuda kernels run 100% of the time, as we eliminate kernel-launch overheads). However, the CPU workload does not profit at all. Given that we are on heterogeneous systems, optimizing GPU workload may actually induce thermal throttling on CPU as CPU will be active more frequently in a given timeframe. Thus, slowdowns can occur in E2E tests.

More performant approaches for models that are too large to fit in vram would be ones where weights are streamed async onto the GPU (e.g. while previous workload is running either on GPU or CPU). Though such approaches are also more complex.

wishstudio committed:
Merge remote-tracking branch 'upstream/master' into cuda_graph_plan

wishstudio committed:
Fix wording.

wishstudio committed:
Fix faulty consecutive updates handling.

wishstudio commented on ggml/src/ggml-cuda/ggml-cuda.cu:
```diff
@@ -2925,220 +2929,255 @@ static bool ggml_cuda_can_fuse(const struct ggml_cgraph * cgraph, int node_idx,
         }
+    } else {
+        cuda_graph->number_consecutive_updates = 0;
     }
```
Good catch!

wishstudio committed:
Fix editorconfig check.

wishstudio commented on ggml/src/ggml-backend.cpp:
```diff
@@ -1413,6 +1454,62 @@ static bool ggml_backend_sched_alloc_splits(ggml_backend_sched_t sched) {
+                if (sched->splits[i].backend_id != sched->plans[i].backend_id) {
+                    create_new_plans = true;
+                    break;
+                }
```
By the way, if the split graph is indeed identical, the CUDA backend will find out via the `is_cuda_graph_update_required` fast path. I don't think it's a good idea to move the checks up to the backend scheduler as different backends could have different definition of **equivalence** depending on various factors as discussed in #14514. We may calculate these factors in the scheduler but the ultimate decision is still better be done in the backend.

wishstudio committed:
Forward declaration for ggml_backend_supports_graph_plan and ggml_backend_supports_graph_plan_update

wishstudio commented:
Added the `ggml_backend_supports_graph_plan` and `ggml_backend_supports_graph_plan_update` APIs to ggml-backend.h to make CI happy.

wishstudio commented:
Looks like there is a nasty destruction order bug. This PR makes backend sched depends on backend for destruction. But in llama_context their order is unspecified. Not sure about the best way to resolve this.

wishstudio commented:
> Looks like there is a nasty destruction order bug. This PR makes backend sched depends on backend for destruction. But in llama_context their order is unspecified. Not sure about the best way to resolve this.

@ggerganov @slaren Would you mind take a look at this matter, as well as this PR in its entirety? Based on my limited knowledge, I think the best way to fix this issue is to create a backend-agnostic interface class for graph plans and move related APIs except `alloc` to this interface, similar to what we currently do with backend buffers. This way won't affect the downstream users of `ggml_backend_sched_ptr`. However, this still requires changes to all backends since graph plan interfaces are part of backend interface. I don't want to rush implementing this without your approval.

wishstudio committed:
Merge branch 'master' into cuda_graph_plan

wishstudio committed:
Fix trailing whitespace
